File: app/Models/Formulaire.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\SoftDeletes;

class Formulaire extends Model
{
    use HasFactory, SoftDeletes;

    /*
    |--------------------------------------------------------------------------
    | Types de champs supportés
    |--------------------------------------------------------------------------
    */
    const TYPES_CHAMPS = [
        'Texte'     => 'Texte court',
        'TexteLong' => 'Texte long',
        'Nombre'    => 'Nombre',
        'Date'      => 'Date',
        'Heure'     => 'Heure',
        'DateHeure' => 'Date et Heure',
        'OuiNon'    => 'Oui / Non',
        'Liste'     => 'Liste déroulante',
        'Checkbox'  => 'Cases à cocher (Checkbox)',
        'Radio'     => 'Boutons radio',
        'Photo'     => 'Photo',
        'Signature' => 'Signature',
        'Document'  => 'Document',
        'GPS'       => 'Coordonnées GPS',
        'QRCode'    => 'QR Code',
        'Materiaux' => 'Matériaux utilisés',
    ];

    // Types nécessitant des choix prédéfinis
    const TYPES_AVEC_CHOIX = ['Liste', 'Checkbox', 'Radio'];

    // Types qui acceptent des fichiers
    const TYPES_FICHIER = ['Photo', 'Signature', 'Document'];

    protected $fillable = [
        'type_intervention_id',
        'nom',
        'description',
        'is_active',
    ];

    protected $casts = [
        'is_active' => 'boolean',
    ];

    /*
    |--------------------------------------------------------------------------
    | Relations
    |--------------------------------------------------------------------------
    */

    // Type d'intervention associé (1 formulaire par type max)
    public function typeIntervention()
    {
        return $this->belongsTo(TypeIntervention::class);
    }

    // Champs/questions du formulaire, ordonnés
    public function questions()
    {
        return $this->hasMany(Question::class)->orderBy('ordre');
    }

    // Questions actives uniquement
    public function questionsActives()
    {
        return $this->hasMany(Question::class)->orderBy('ordre');
    }
}

File: mobile_technicien/lib/features/formulaire/domain/entities/formulaire.dart
import 'package:equatable/equatable.dart';

/// Miroir de app/Models/Formulaire.php — uniquement les champs lus par l'app.
/// `Formulaire::TYPES_CHAMPS` définit les valeurs exactes de [Question.type] :
/// Texte, TexteLong, Nombre, Date, Heure, DateHeure, OuiNon, Liste, Checkbox,
/// Radio, Photo, Signature, Document, GPS, QRCode, Materiaux (ce dernier est
/// ignoré ici — géré par les endpoints matériaux dédiés, cf.
/// RemplissageFormulaireService::sauvegarderReponses qui le skip aussi).
class Formulaire extends Equatable {
  final int id;
  final String nom;
  final String? description;
  final List<Question> questions;

  const Formulaire({required this.id, required this.nom, required this.description, required this.questions});

  @override
  List<Object?> get props => [id, nom, description, questions];
}

class Question extends Equatable {
  final int id;
  final String question;
  final String type;
  final bool obligatoire;
  final int ordre;
  final String? placeholder;
  final String? valeurParDefaut;
  final double? nombreMin;
  final double? nombreMax;
  final String? nombreUnite;
  final int fichiersMax;
  final List<ChoixQuestion> choix;

  const Question({
    required this.id,
    required this.question,
    required this.type,
    required this.obligatoire,
    required this.ordre,
    required this.placeholder,
    required this.valeurParDefaut,
    required this.nombreMin,
    required this.nombreMax,
    required this.nombreUnite,
    required this.fichiersMax,
    required this.choix,
  });

  static const _typesAvecChoix = ['Liste', 'Checkbox', 'Radio'];
  static const _typesFichier = ['Photo', 'Signature', 'Document'];

  bool get necessiteChoix => _typesAvecChoix.contains(type);

  bool get estTypeFichier => _typesFichier.contains(type);

  @override
  List<Object?> get props => [id, question, type, obligatoire, ordre, choix];
}

class ChoixQuestion extends Equatable {
  final int id;
  final String libelle;
  final String valeur;

  const ChoixQuestion({required this.id, required this.libelle, required this.valeur});

  @override
  List<Object?> get props => [id, libelle, valeur];
}
